src/Dictionaries/CacheDictionary.cpp

```diff
@@ -276,19 +276,81 @@ void CacheDictionary::getString(
     getItemsString(attribute, ids, out, [&](const size_t) { return StringRef{def}; });
 }
 
+template<class... Ts>
+struct Overloaded : Ts... {using Ts::operator()...;};
 
-/// returns cell_idx (always valid for replacing), 'cell is valid' flag, 'cell is outdated' flag
+template<class... Ts>
+Overloaded(Ts...) -> Overloaded<Ts...>;
+
+std::string CacheDictionary::AttributeValuesForKey::dump()
+{
+    std::ostringstream os;
+    for (auto & attr : values)
+        std::visit(Overloaded {
+            [&os](UInt8 arg)   { os << "type: UInt8, value: "   <<  std::to_string(arg) << "\n"; },
+            [&os](UInt16 arg)  { os << "type: UInt16, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](UInt32 arg)  { os << "type: UInt32, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](UInt64 arg)  { os << "type: UInt64, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](UInt128 arg) { os << "type: UInt128, value: " << arg.toHexString() << "\n"; },
+            [&os](Int8 arg)   { os << "type: Int8, value: "   <<  std::to_string(arg) << "\n"; },
+            [&os](Int16 arg)  { os << "type: Int16, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](Int32 arg)  { os << "type: Int32, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](Int64 arg)  { os << "type: Int64, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](Decimal32 arg)   { os << "type: Decimal32, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](Decimal64 arg)   { os << "type: Decimal64, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](Decimal128)  { os << "type: Decimal128, value: ???" << "\n" ; },
+            [&os](Float32 arg)   { os << "type: Float32, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](Float64 arg)   { os << "type: Float64, value: "  <<  std::to_string(arg) << "\n"; },
+            [&os](String arg)  { os << "type: String, value: " <<  arg + "\n"; }
+        }, attr);
+    return os.str();
+};
+
+
+std::string CacheDictionary::UpdateUnit::dumpFoundIds()
+{
+    std::ostringstream os;
+    for (auto it : found_ids)
+    {
+        os << "Key: " << std::to_string(it.first) << "\n";
+        if (it.second.found)
+            os << it.second.dump() << "\n";
+    }
+    return os.str();
+};
+
+/// Returns cell_idx in handmade open addressing cache table.
 /// true  false   found and valid
-/// false true    not found (something outdated, maybe our cell)
+/// false true    found, but expired
 /// false false   not found (other id stored with valid data)
 /// true  true    impossible
-///
-/// todo: split this func to two: find_for_get and find_for_set
-CacheDictionary::FindResult CacheDictionary::findCellIdx(const Key & id, const CellMetadata::time_point_t now) const
+CacheDictionary::FindResult CacheDictionary::findCellIdxForGet(const Key & id, const time_point_t now) const
```

Comment:
A state enum would probably be more appropriate than two booleans. Btw, in relatively modern C++ you can return and destructure tuples, e.g. `auto [value, state] = findCellIdx()`. I think it even works in Arcadia builds.

Reply:
Done.